parutils/src/lib.rs

```diff
@@ -1,4 +1,185 @@
-#![cfg_attr(feature = "strict", deny(warnings))]
+use std::convert::Infallible;
+use std::future::Future;
+use std::sync::Arc;
 
-mod parallel_utils;
-pub use parallel_utils::*;
+use thiserror::Error;
+use tokio::sync::{AcquireError, Semaphore};
+use tokio::task::{JoinError, JoinSet};
+
+#[derive(Debug, Error)]
+pub enum ParutilsError<E: Send + Sync + 'static> {
+    #[error(transparent)]
+    Join(#[from] JoinError),
+    #[error(transparent)]
+    Acquire(#[from] AcquireError),
+    #[error(transparent)]
+    Task(E),
+}
+
+/// Runs all the futures in the provided iterator with a maximum concurrency limit.
+///
+/// This function requires that a permit is acquired from the provided semaphore before any work
+/// is done for a future, thus limiting concurrency based on the number of permits in the semaphore.
+///
+/// Each future in the iterator must return a `Result<T, E>`. If any future returns an error,
+/// or if there is a `JoinError` or failure to acquire a semaphore permit, the function will
+/// return an error as soon as possible.
+///
+/// If all tasks complete successfully, the function returns a `Vec<T>` containing the results
+/// of the successful futures, in the same order as they were produced by the iterator.
+///
+/// # Arguments
+///
+/// * `futures_it` - An iterator of futures, where each future resolves to a `Result<T, E>`.
+/// * `max_concurrent` - An `Arc<Semaphore>` that limits the number of concurrent tasks.
+///
+/// # Type Parameters
+///
+/// * `Fut` - The type of the futures in the iterator. Each future must output a `Result<T, E>`.
+/// * `T` - The type of the successful result produced by each future.
+/// * `E` - The type of the error produced by each future.
+///
+/// # Returns
+///
+/// A `Result` containing:
+/// * `Ok(Vec<T>)` - A vector of successful results if all tasks complete successfully.
+/// * `Err(ParutilsError<E>)` - An error if any task fails, a semaphore permit cannot be acquired, or a `JoinError`
+///   occurs.
+///
+/// # Errors
+///
+/// This function returns a `ParutilsError<E>` in the following cases:
+/// * A task returns an error of type `E`.
+/// * A semaphore permit cannot be acquired.
+/// * A `JoinError` occurs while waiting for a task to complete.
+///
+/// # Example
+///
+/// ```rust
+/// use std::sync::Arc;
+///
+/// use parutils::tokio_run_max_concurrency_fold_result_with_semaphore;
+/// use tokio::sync::Semaphore;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let semaphore = Arc::new(Semaphore::new(2)); // Limit concurrency to 2 tasks.
+///     let futures = (1..=3).map(|n| async move { Ok::<_, ()>(n) });
+///
+///     let results =
+///         tokio_run_max_concurrency_fold_result_with_semaphore(futures.into_iter(), semaphore)
+///             .await;
+///     assert_eq!(results.unwrap(), vec![1, 2, 3]);
+/// }
+/// ```
+pub async fn tokio_run_max_concurrency_fold_result_with_semaphore<Fut, T, E>(
```

Comment:
nit: IMO, the `tokio_run_max_concurrency*` name is a bit lengthy. We could probably rename to just `run_constrained_*` instead?